gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp

```diff
@@ -132,35 +162,34 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
       // If number of queries is high enough do it here
       else
       {
-        std::cout << " GPU: ";
+        PCL_DEBUG(" GPU: ");
         // Copy buffer
         queries_device = DeviceArray<PointXYZ>(queries_device_buffer.ptr(),queries_host.size());
         // Move queries to GPU
         queries_device.upload(queries_host);
         // Execute search
         tree->radiusSearch(queries_device, tolerance, max_answers, result_device);
         // Copy results from GPU to Host
-        result_device.sizes.download (sizes);
-        result_device.data.download (data);
+        result_device.sizes.download(sizes);
+        gpu_indices.clear();
+        pcl::detail::economical_download(result_device, sizes, max_answers, gpu_indices);
```

Comment:
I'm wondering if a function is the right fit, instead of a lambda with capture. Thoughts?

Pros:
* capture the variables properly, so the call syntax is less noisy
    ```
    economical_download(result_device, max_answers, gpu_indices);  // with lambda that captures other variables
    economical_download(result_device, sizes, max_answers, gpu_indices, gpu_tmp);  // with function
* No additional symbol added under detail namespace

Cons:
* lambda will be placed around L100, and I've found others complaining about readability differences between free functions and non-immediately used lambdas
* lambda might have to be mutable (for `gpu_tmp`)

Reply:
Thank you so much for the comment, @kunaltyagi! I usually appreciate the lambda's conciseness and happily thought about its use here. I modified the program as follows:

```c++
  pcl::Indices sizes, data, cpu_tmp, gpu_tmp, gpu_indices;
  auto economical_download = [&](const auto& source_indices,
        const auto& buffer_indices, auto& downloaded_indices) {
    for (std::size_t qp = 0; qp < buffer_indices.size(); qp++) {
        std::size_t begin = qp * max_answers;
        const int* const pdata = source_indices.data.ptr() + begin;
        gpu_tmp.resize(buffer_indices[qp]);
        const std::size_t bytes = (buffer_indices[qp]) * sizeof(int);
        cudaMemcpy(&gpu_tmp[0], pdata, bytes, cudaMemcpyDeviceToHost);
        cudaDeviceSynchronize();
        downloaded_indices.insert(downloaded_indices.end(), gpu_tmp.begin(), gpu_tmp.end());
    }
  ;};
```
The function now has one fewer argument (`max_answers`), and `gpu_tmp` gets captured. I very much appreciate these two aspects. Did you have another idea for writing the lambda?

Concluding, I am undecided on whether introducing the lambda is beneficial. The lambda is a bit long and interrupts the code's flow for my taste. I would suggest using either lambda and `gpu_tmp` or continue using the function and use `gpu_tmp` as a local variable. The allocations do not have a substantial impact, and the vector gets used only locally. Do you have a preference?

Reply:
Yeah, that would have been my lambda as well.

> The allocations do not have a substantial impact, and the vector gets used only locally.

This is curious. Assuming buffer_indices are not all same, the resize would incur a cost right before a CPU write. Perhaps my intuition is wrong.

The current speed-up is anyways quite good :)

Reply:
Great to see we would have written identical lambdas! And yes, oh boy, these allocations are playing tricks with me all day long! 

However, the [most recent flamegraph](https://user-images.githubusercontent.com/10357496/113475210-6f56aa80-9474-11eb-99a3-28087dfd4bd8.png) documents that allocations are not substantial. The right end of the image shows a yellow vertical bar named `pcl::gpu::EuclidianClust`.  The bar's length is proportional to the time spent inside this function, and vertical bars stacked above denote the function's callees. As shown, `economical_download` is predominantly waiting for Cuda-related calls to complete. The small vertical slice whose name is too small to be read on this image contains code for the vector memory allocations. In contrast, the [first flamegraph](https://user-images.githubusercontent.com/10357496/112655136-63556200-8e50-11eb-8479-a832527e09d9.jpg) highlights that the unoptimized program waited a long time for `std::vector::resize` to complete.

Reply:
Then no point in optimizing this :)